runtime/compiler/optimizer/MethodHandleTransformer.cpp

```diff
@@ -0,0 +1,562 @@
+              {
+              TR::VMAccessCriticalSection dereferenceKnownObjectField(comp()->fej9());
+              uintptr_t mhObject = knot->getPointer(mhIndex);
+              uintptr_t mnObject = comp()->fej9()->getReferenceField(mhObject, "initMethod", "Ljava/lang/invoke/MemberName;");
+              auto mnIndex = knot->getOrCreateIndex(mnObject);
+              if (trace())
+                 traceMsg(comp(), "Get DirectMethodHandle.initMethod known object %d\n", mnIndex);
+              return mnIndex;
+              }
+           }
+        }
+      }
+
+   return UNKNOWN;
+   }
+
+// Store to local variable will change object info
+// Update _currentObjectInfo
+void
+TR_MethodHandleTransformer::visitStoreToLocalVariable(TR::TreeTop* tt, TR::Node* node)
+   {
+   TR::Node *rhs = node->getFirstChild();
+   TR::Symbol *local = node->getSymbolReference()->getSymbol();
+   if (rhs->getDataType().isAddress())
+      {
+      // Get object info of the rhs
+      LocalObjectInfo newObject = getObjectInfoOfNode(rhs);
+      if (trace())
+         traceMsg(comp(), "rhs of store n%dn is obj%d\n", node->getGlobalIndex(), newObject);
+
+      if (newObject != UNUSED)
+         {
+         if (!_currentObjectInfo)
+            _currentObjectInfo = new (comp()->trMemory()->currentStackRegion()) ObjectInfo(_numLocals, UNUSED, comp()->trMemory()->currentStackRegion());
+         if (trace())
+            {
+            LocalObjectInfo oldObject = (*_currentObjectInfo)[local->getLocalIndex()];
+            traceMsg(comp(), "Local #%2d obj%d -> obj%d at node n%dn\n", local->getLocalIndex(), oldObject, newObject,  node->getGlobalIndex());
+            }
+         (*_currentObjectInfo)[local->getLocalIndex()] = newObject;
+         }
+      }
+   }
+
+// Visit indirect load, fold the load if applicable
+void TR_MethodHandleTransformer::visitIndirectLoad(TR::TreeTop* tt, TR::Node* node)
+   {
+   auto symRef = node->getSymbolReference();
+   if (symRef->hasKnownObjectIndex())
+      {
+      if (trace())
+         traceMsg(comp(), "Indirect load n%dn is obj%d\n", node->getGlobalIndex(), symRef->getKnownObjectIndex());
+      return;
+      }
+
+   auto symbol = node->getSymbol();
+   if (!symRef->isUnresolved() && symbol &&
+       (symbol->isFinal() || symbol->isArrayShadowSymbol()))
+      {
+      auto baseNode = symbol->isArrayShadowSymbol() ? node->getFirstChild()->getFirstChild() : node->getFirstChild();
+      if (!baseNode->getOpCode().hasSymbolReference() ||
+          baseNode->hasUnresolvedSymbolReference())
+         return;
+
+      auto baseSymRef = baseNode->getSymbolReference();
+      LocalObjectInfo baseObj = getObjectInfoOfNode(baseNode);
+      if (trace())
+         traceMsg(comp(), "base object for indirect load n%dn is obj%d\n", node->getGlobalIndex(), baseObj);
+
+      auto knot = comp()->getKnownObjectTable();
+      if (isKnownObject(baseObj)  && !knot->isNull(baseObj))
```

Comment:
We need a test on `knot != NULL` because knot can be disabled through an option, though it could be argues that `isKnownObject(baseObj)` cannot return `true` if knot doesn't exist.